### PR TITLE
instrument aerospike v4, v5

### PR DESCRIPTION
**Pull request by khanayan123**
### What does this PR do?
instruments aerospike v4 & v5 node js client library by instrumenting the command class (https://github.com/aerospike/aerospike-client-nodejs/blob/master/lib/commands/command.js), which is what every operation of interest goes through.

### Motivation
feature request

### Additional Notes
the span attributes try to follow the convention laid out by the dotnet tracer (https://github.com/datadog/dd-trace-dotnet/blob/6aab5e1b02bec9c9b68a33cd06cc9e7a774f14de/tracer/test/snapshots/AerospikeTests.SchemaV0.verified.txt#L1) with some special handling for the query and index commands

**Side note: I'm currently testing aerospike with node 14 in this PR, but once it gets approved i'll remove the test for it from the CI and revert the package.json. After merging the pr, node 14 will get tested in the v3 release branch

Also here is the java instrumentation for reference: https://github.com/DataDog/dd-trace-java/blob/78c55e8b1b111a4ff5abc7374cf7f28606f8e1ae/dd-java-agent/instrumentation/aerospike-4/src/main/java/datadog/trace/instrumentation/aerospike4/AerospikeClientDecorator.java

### CODE REVIEW

File: package.json
Comment (tlhunter) [inline package.json:68]: Was this added on accident?

Comment (khanayan123) [reply package.json:68]: I'm currently testing aerospike with node 14 in this PR, but once it gets approved i'll remove the test for it from the CI and revert the package.json. After merging the pr, node 14 will get tested in the v3 release branch

File: packages/datadog-instrumentations/src/aerospike.js
Comment (tlhunter) [inline packages/datadog-instrumentations/src/aerospike.js:77]: Can you rewrite all these that use `<` and `>` to be `<=` and `>=`? I find it more intuitive if the ranges are "inclusive".

File: scripts/install_plugin_modules.js
Comment (tlhunter) [inline scripts/install_plugin_modules.js:133]: This exception just for one package looks interesting. What's the story here?

Comment (khanayan123) [reply scripts/install_plugin_modules.js:133]: `Command: npm install @mapbox/node-pre-gyp
Arguments: 
Directory: /home/runner/work/dd-trace-js/dd-trace-js/versions/aerospike/node_modules/aerospike
Output:
TypeError: workspaces config expects an Array`

Running yarn services, was giving me this error^. Tried a bunch of things to get it to work, & this suggestion made it work:

https://stackoverflow.com/questions/71747375/migrating-to-yarn-3-workspaces-nohoist

Comment (rochdev) [reply scripts/install_plugin_modules.js:133]: What's the harm of always providing both options?

File: .github/workflows/plugins.yml
Comment (rochdev) [inline .github/workflows/plugins.yml:38]: Why not just use `oldest` and `latest` like other plugins?

File: package.json
Comment (rochdev) [reply package.json:68]: Then the PR should be a draft, otherwise it could be merged by accident.

File: packages/datadog-instrumentations/src/aerospike.js
Comment (rochdev) [inline packages/datadog-instrumentations/src/aerospike.js:41]: Why is the long form used here for the channels? Doesn't `TracingChannel` support this use case perfectly?

File: packages/datadog-instrumentations/src/aerospike.js
Comment (rochdev) [inline packages/datadog-instrumentations/src/aerospike.js:71]: This check makes it only for Node 20 exactly and not "up until 20".

File: packages/datadog-instrumentations/src/aerospike.js
Comment (rochdev) [reply packages/datadog-instrumentations/src/aerospike.js:77]: What happens in Node `>20` anyway? It doesn't look like any instrumentation provides supports for that.

File: packages/datadog-plugin-aerospike/src/index.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/src/index.js:20]: If these are scoped to the condition branch, let's use `const` instead to avoid confusion.

File: packages/datadog-plugin-aerospike/src/index.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/src/index.js:22]: Commands don't usually start uppercase for other plugins. Does this match how it's done in .NET?

File: packages/datadog-plugin-aerospike/src/index.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/src/index.js:33]: Is this an actual type supported by our backend?

File: packages/datadog-plugin-aerospike/src/index.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/src/index.js:12]: Should the arguments be normalized on the instrumentation side? My worry is that not only commands clearly have different arguments, but they could also differ between versions.

File: packages/datadog-plugin-aerospike/src/index.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/src/index.js:46]: Let's add checks that `currentStore` is set before trying to access `span` to avoid issues when external subscribers are present.

Comment (rochdev) [reply packages/datadog-plugin-aerospike/src/index.js:46]: Also, let's keep the name of the variable consistent to make it obvious it's the same thing, so if it's usually called `ctx` let's call it `ctx` everywhere.

File: packages/datadog-plugin-aerospike/src/index.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/src/index.js:58]: Similar to the above, let's make sure this even exists.

File: packages/datadog-plugin-aerospike/src/index.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/src/index.js:66]: Are these tags specified anywhere? And if not, do they match existing implementations?

File: packages/datadog-plugin-aerospike/test/index.spec.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/test/index.spec.js:59]: Shouldn't this be using the naming schema?

File: packages/datadog-plugin-aerospike/test/index.spec.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/test/index.spec.js:77]: This one is interesting because it might happen outside of a request. Is it explicitly sent by the user or is it an implicit command sent automatically when other types of commands are sent?

**Review by rochdev — CHANGES_REQUESTED** [review]
I left a bunch of comments but it's mostly questions and small things. Overall the PR LGTM.

File: packages/datadog-instrumentations/src/aerospike.js
Comment (khanayan123) [reply packages/datadog-instrumentations/src/aerospike.js:77]: just can't install aerospike with node >20

File: scripts/install_plugin_modules.js
Comment (khanayan123) [reply scripts/install_plugin_modules.js:133]: then aerospike specifically won't build because it has this set:

pkg.workspaces = {
      nohoist: ['**/**']
  }

File: .github/workflows/plugins.yml
Comment (khanayan123) [reply .github/workflows/plugins.yml:38]: latest will test > node 20, will change it to oldest once the pr's approved, it's mainly for my own testing since older aerospike versions don't work on arm

File: packages/datadog-instrumentations/src/aerospike.js
Comment (khanayan123) [reply packages/datadog-instrumentations/src/aerospike.js:41]: except for the explicit error handling, where you have to check if the callback's first arg is not null which means there was an error and so it needs to be published to the subscriber

File: packages/datadog-plugin-aerospike/src/index.js
Comment (khanayan123) [reply packages/datadog-plugin-aerospike/src/index.js:22]: yes .NET uses upper cases

Comment (khanayan123) [reply packages/datadog-plugin-aerospike/src/index.js:22]: see here: https://github.com/datadog/dd-trace-dotnet/blob/6aab5e1b02bec9c9b68a33cd06cc9e7a774f14de/tracer/test/snapshots/AerospikeTests.SchemaV0.verified.txt#L1

File: packages/datadog-plugin-aerospike/test/index.spec.js
Comment (khanayan123) [reply packages/datadog-plugin-aerospike/test/index.spec.js:77]: Well I believe it is always sent because aerospike works by chaining callbacks and promises, so you would first always have to connect and then on the promise/callback of that call you can perform other requests inside. But no I don't think you connect first and then later on run a request, everything has to be chained in the connects call back

File: scripts/install_plugin_modules.js
Comment (rochdev) [reply scripts/install_plugin_modules.js:133]: Aren't the 2 options equivalent? I'm not sure I understand the explanation in the above link.

File: packages/datadog-instrumentations/src/aerospike.js
Comment (rochdev) [reply packages/datadog-instrumentations/src/aerospike.js:41]: TracingChannel does that: https://github.com/DataDog/dc-polyfill/blob/56954afcd6c7f2dbabee30d861c89d52e61a1d3d/patch-tracing-channel.js#L136

File: packages/datadog-instrumentations/src/aerospike.js
Comment (Qard) [inline packages/datadog-instrumentations/src/aerospike.js:49]: Please give this a name. The purpose of `args[0]` is very non-obvious.

File: packages/datadog-instrumentations/src/aerospike.js
Comment (Qard) [inline packages/datadog-instrumentations/src/aerospike.js:44]: Is there a reason to not just add actual parameters so you can have a more descriptive reference name here?

```suggestion
          ctx.error = error
```

File: packages/datadog-instrumentations/src/aerospike.js
Comment (Qard) [inline packages/datadog-instrumentations/src/aerospike.js:61]: I would prefer we avoid single letter variable names. At the _least_ I would make this `err` to be a bit more clear about what it is.

File: packages/datadog-plugin-aerospike/src/index.js
Comment (Qard) [inline packages/datadog-plugin-aerospike/src/index.js:24]: Why rename these to something else only to map them back in the object construction?

```suggestion
      const { ns, set } = commandArgs[2]
      meta = getMetaForQuery({ ns, set })
```

File: packages/datadog-plugin-aerospike/test/index.spec.js
Comment (Qard) [inline packages/datadog-plugin-aerospike/test/index.spec.js:57]: This is only used in one place. Could just embed it in-place. But also, it shouldn't use backticks if it's a static string like this.

File: packages/datadog-plugin-aerospike/test/index.spec.js
Comment (Qard) [inline packages/datadog-plugin-aerospike/test/index.spec.js:212]:
```suggestion
                  const queryPolicy = {
                    filterExpression: exp.keyExist('uniqueExpKey')
                  }
```

**Review by Qard — COMMENTED** [review]
Commented with a few nits, but otherwise LGTM once the v14 testing bits are reverted to make this landable.

File: scripts/install_plugin_modules.js
Comment (khanayan123) [reply scripts/install_plugin_modules.js:133]: Technically they are, but for some reason aerospike can't build with the yarn services command if this is set pkg.workspaces = {
nohoist: ['/']
}

File: packages/datadog-plugin-aerospike/src/index.js
Comment (khanayan123) [reply packages/datadog-plugin-aerospike/src/index.js:66]: got it from the dotnet snapshots, since there is no spec for aerospike i believe both dotnet and java do something a little different

Comment (khanayan123) [reply packages/datadog-plugin-aerospike/src/index.js:66]: dotnet snapshots

https://github.com/datadog/dd-trace-dotnet/blob/6aab5e1b02bec9c9b68a33cd06cc9e7a774f14de/tracer/test/snapshots/AerospikeTests.SchemaV0.verified.txt#L1

File: packages/datadog-plugin-aerospike/src/index.js
Comment (khanayan123) [reply packages/datadog-plugin-aerospike/src/index.js:33]: don't know for sure, but I know both the dotnet and java tracers add it on their traces

File: packages/datadog-instrumentations/src/aerospike.js
Comment (khanayan123) [reply packages/datadog-instrumentations/src/aerospike.js:77]: maybe a newer version of aerospike adds support for node >20 but the latest aerospike release can't be installed on node 20

File: packages/datadog-plugin-aerospike/src/index.js
Comment (khanayan123) [reply packages/datadog-plugin-aerospike/src/index.js:12]: shouldn't they be normalized on the subscriber side? since the instrumentation is just attaching every argument the command gets to the ctx & then the subscriber decides what to do with that?

Comment (rochdev) [reply packages/datadog-plugin-aerospike/src/index.js:15]: What I meant is that the relevant information could be extracted from the publisher into a normalized data structure that contains the information in a way that is guaranteed not to change between versions. This has the benefit of the subscriber never having to worry about changes in the arguments, and having to handle the same normalization logic in every subscriber.

However, since both the publisher and subscriber lives in the same repo, it might not be worth it to tackle this right now.

File: scripts/install_plugin_modules.js
Comment (rochdev) [reply scripts/install_plugin_modules.js:133]: Ok, we can always figure out this one later.

File: packages/datadog-plugin-aerospike/test/naming.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/test/naming.js:11]: For databases this should be only the service name in v1, so `test`.

File: packages/datadog-plugin-aerospike/src/index.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/src/index.js:21]: Small nit, but this whole condition could be moved to a `getMeta` function.

File: packages/datadog-instrumentations/src/aerospike.js
Comment (rochdev) [reply packages/datadog-instrumentations/src/aerospike.js:42]: As discussed offline, let's remove these completely from the instumentation and instead focus on the test strategy for the Node versions.

File: packages/datadog-plugin-aerospike/src/index.js
Comment (jbertran) [inline packages/datadog-plugin-aerospike/src/index.js:13]: This should also report `peer.service` - it seems like the relevant precursor is `aerospike.namespace`.

All you need to do (see for ex. `packages/datadog-plugin-cassandra-driver/src/index.js`) is define the `peerServicePrecursors` static getter to return `['aerospike.namespace']`, and test with the `withPeerService` fixture.

File: packages/datadog-plugin-aerospike/test/index.spec.js
Comment (jbertran) [inline packages/datadog-plugin-aerospike/test/index.spec.js:265]: Can we have a `withNamingSchema` check without configuration to verify that we're correctly falling back onto whatever the tracer service is when we don't have more specific configuration?

File: packages/dd-trace/src/service-naming/schemas/v0/storage.js
Comment (rochdev) [reply packages/dd-trace/src/service-naming/schemas/v0/storage.js:43]: I disagree. It's more important to have consistent behaviour for any user that is unaware of the schema versions.

Comment (jbertran) [reply packages/dd-trace/src/service-naming/schemas/v0/storage.js]: I don't have a strong opinion either way, but having the future service name in `v0` would lighten the user's load when moving to the better `v1` wrt existing monitors etc.

**Review by jbertran — APPROVED** [review]

File: packages/datadog-instrumentations/src/aerospike.js
Comment (rochdev) [inline packages/datadog-instrumentations/src/aerospike.js:40]: This code is polluting the instrumentation for the sake of testing. This should only be done in tests. Also, doesn't the `RANGE` environment variable already support this use case?

File: packages/datadog-plugin-aerospike/test/naming.js
Comment (rochdev) [inline packages/datadog-plugin-aerospike/test/naming.js:7]: As discussed, let's focus on consistent behaviour. The whole point of v1 is to fix the problem in a consistent manner, but in the meantime, v0 services should be consistent with each other.

**Review by rochdev — CHANGES_REQUESTED** [review]

**Review by rochdev — DISMISSED** [review]

**Review by rochdev — APPROVED** [review]

**Review by rochdev — APPROVED** [review]

**Review by jbertran — APPROVED** [review]

**Pull request merged**
